tools/nvprefs/nvprefs.cpp

```diff
@@ -0,0 +1,227 @@
+      fill_nvapi_string(application.userFriendlyName, sunshine_application_path);
+      fill_nvapi_string(application.launcher, L"");
+
+      status = NvAPI_DRS_CreateApplication(session_handle, profile_handle, &application);
+      if (status != NVAPI_OK) return status;
+      session_modified = true;
+    }
+
+    NVDRS_SETTING setting = {};
+    setting.version = NVDRS_SETTING_VER1;
+    status = NvAPI_DRS_GetSetting(session_handle, profile_handle, PREFERRED_PSTATE_ID, &setting);
+
+    if (status != NVAPI_OK ||
+        setting.settingLocation != NVDRS_CURRENT_PROFILE_LOCATION ||
+        setting.u32CurrentValue != PREFERRED_PSTATE_PREFER_MAX) {  // Set power setting if needed
+      setting = {};
+      setting.version = NVDRS_SETTING_VER1;
+      setting.settingId = PREFERRED_PSTATE_ID;
+      setting.settingType = NVDRS_DWORD_TYPE;
+      setting.settingLocation = NVDRS_CURRENT_PROFILE_LOCATION;
+      setting.u32CurrentValue = PREFERRED_PSTATE_PREFER_MAX;
+
+      status = NvAPI_DRS_SetSetting(session_handle, profile_handle, &setting);
+      if (status != NVAPI_OK) return status;
+      session_modified = true;
+
+      std::cout << "Modified application power profile" << std::endl;
+    }
+    else {
+      std::cout << "Application power profile has already been modified" << std::endl;
+    }
+
+    return NVAPI_OK;
+  }
+
+  NvAPI_Status
+  check_and_modify_global_profile(NvDRSSessionHandle session_handle, bool &session_modified) {
+    NvAPI_Status status;
+
+    NvDRSProfileHandle profile_handle = 0;
+    status = NvAPI_DRS_GetBaseProfile(session_handle, &profile_handle);
+    if (status != NVAPI_OK) return status;
+
+    NVDRS_SETTING setting = {};
+    setting.version = NVDRS_SETTING_VER1;
+    status = NvAPI_DRS_GetSetting(session_handle, profile_handle, OGL_CPL_PREFER_DXPRESENT_ID, &setting);
+
+    if (status != NVAPI_OK ||
+        setting.u32CurrentValue != OGL_CPL_PREFER_DXPRESENT_PREFER_ENABLED) {  // Set OpenGL/Vulkan DXGI swapchain setting if needed
+      setting = {};
+      setting.version = NVDRS_SETTING_VER1;
+      setting.settingId = OGL_CPL_PREFER_DXPRESENT_ID;
+      setting.settingType = NVDRS_DWORD_TYPE;
+      setting.settingLocation = NVDRS_CURRENT_PROFILE_LOCATION;
+      setting.u32CurrentValue = OGL_CPL_PREFER_DXPRESENT_PREFER_ENABLED;
+
+      status = NvAPI_DRS_SetSetting(session_handle, profile_handle, &setting);
+      if (status != NVAPI_OK) return status;
+      session_modified = true;
+
+      std::cout << "Modified global OpenGL/Vulkan DXGI swapchain" << std::endl;
+    }
+    else {
+      std::cout << "No need to modify OpenGL/Vulkan DXGI swapchain" << std::endl;
+    }
+
+    return NVAPI_OK;
+  }
+
+  NvAPI_Status
+  restore_global_profile(NvDRSSessionHandle session_handle, bool &session_modified) {
+    NvAPI_Status status;
+
+    NvDRSProfileHandle profile_handle = 0;
+    status = NvAPI_DRS_GetBaseProfile(session_handle, &profile_handle);

```

Comment:
You might want to use [`NvAPI_DRS_GetCurrentGlobalProfile()`](https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__drsapi.html#ga68dbe5987a8f530a0041ba94019b0ec0) instead. See [here](https://github.com/Lucide/vsync-tool#nvdrs_settingsettinglocation-values).

Reply:
From [NVIDIA Driver Settings Programming Guide](https://developer.download.nvidia.com/tools/docs/PG-5116-001_v02_public.pdf):
> A profile with one or more applications is called an Application Profile. Applications
that have a profile have their settings applied automatically when the application starts.
> 
> The Base Profile is a profile that always exists and is not associated with a specific
application. The settings in the Base Profile are applied to all processes on the system
automatically.
> 
> A profile with no applications is called a Global Profile. The settings from a Global
Profile are applied to all processes on the system, but only if that profile is selected to be
the Current Global Profile in the system.

...

> When an application starts, there might be three profiles contesting to provide settings to
the NVIDIA driver: the Application Profile, the Current Global Profile and the Base
Profile.
> 
> The settings from the Application Profile have the highest priority and override any
settings on the other two profiles. Next in priority order comes the Current Global
Profile, and finally the Base Profile.

No idea if anyone actually uses global profiles, but if somebody changes the profile we will lose the setting, even if the profile doesn't change that setting to begin with. There might be some sense in changing both base and global profiles at the same time, but it's going to complicate the undo logic, I will think about it.

Reply:
`NVDRS_GLOBAL_PROFILE_LOCATION` points to the active global profile. In most cases, yes, it will be the base profile. If the user uses something like [NVIDIA Profile Inspector](https://github.com/Orbmu2k/nvidiaProfileInspector), that might not be the case. You do not need to keep track of both, only the global profile will be sufficient.

Reply:
Global profile doesn't replace base profile, it extends it. It can't "point" to base profile. Similarly to application profile. Basically both global and application profiles work on opt-in basis and only override specific settings.

And regarding NVIDIA Profile Inspector, I don't see the ability to *change* global profile. It can change base profile, create named profile, and add applications to it.

Reply:
I think I remember that, in a "standard" situation, a global profile might not exist. In this case, `NvAPI_DRS_GetCurrentGlobalProfile()` returns the same as `NvAPI_DRS_GetBaseProfile()`, and `.settingLocation` is always `NVDRS_GLOBAL_PROFILE_LOCATION`. A different current global profile can be set with [`NvAPI_DRS_SetCurrentGlobalProfile()`](https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__drsapi.html#ga35f52bf2ffa9d2a8f758912c41f4e936). You're right, in the profile inspector, you can't set the current global profile, but if you do programmatically, it will reflect the change.

Reply:
Sorry for seemingly ignoring you, just had to basically rewrite a lot of stuff and this question was hardly a priority at that moment.

After giving it more thought, I think the most logical way is to follow what NVIDIA control panel itself does. Whether after `NvAPI_DRS_SetCurrentGlobalProfile()` it switches to that profile, or sticks to the base one. Will test this and possibly make modifications accordingly.

Reply:
No worries, mine were just suggestions. If the code is sound, corrections will be trivial should issues arise.

Reply:
Just tested, NVIDIA control panel only displays and modifies base profile even after custom one is set, and that custom profile properly inherits from the base one. So unless custom global profile explicitly overrides the settings we change we should be fine, and I'd rather not change settings that are essentially hidden from the user. Sticking to the base profile for now, unless some particular use-case arises where it is not enough.